chisel/src/result.rs

```diff
@@ -0,0 +1,276 @@
+    Hex(Module, PathBuf),
+    #[cfg(feature = "wabt")]
+    Wat(Module, PathBuf),
+}
+
+#[derive(Clone)]
+/// Individual module execution result. Left-hand field is the module name, and left-hand is the
+/// return value.
+pub enum ModuleResult {
+    Creator(String, Result<bool, ModuleError>),
+    Translator(String, Result<bool, ModuleError>),
+    Validator(String, Result<bool, ModuleError>),
+}
+
+impl ChiselResult {
+    pub fn new() -> Self {
+        ChiselResult(Vec::new())
+    }
+
+    pub fn rulesets_mut(&mut self) -> &mut Vec<RulesetResult> {
+        &mut self.0
+    }
+
+    pub fn rulesets(&self) -> &Vec<RulesetResult> {
+        &self.0
+    }
+}
+
+impl RulesetResult {
+    pub fn new(name: String) -> Self {
+        RulesetResult {
+            ruleset_name: name,
+            results: Vec::new(),
+            output_path: PathBuf::new(),
+            output_module: None,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        self.ruleset_name.as_str()
+    }
+
+    pub fn results_mut(&mut self) -> &mut Vec<ModuleResult> {
+        &mut self.results
+    }
+
+    pub fn set_output_path(&mut self, path: PathBuf) {
+        self.output_path = path;
+    }
+
+    pub fn set_output_module(&mut self, module: Module) {
+        self.output_module = Some(module);
+    }
+
+    /// Write output module to specified file if the module was mutated.
+    /// Returns Ok(false) if there is no mutation.
+    /// Returns error on writer error or invalid mode.
+    pub fn write(&mut self, mode: &str) -> Result<bool, Box<dyn Error>> {
+        if let Some(module) = self.output_module.take() {
+            let writer = match mode {
+                "bin" => Some(ChiselOutputWriter::Bin(
+                    module,
+                    PathBuf::from(&self.output_path),
+                )),
+                "hex" => Some(ChiselOutputWriter::Hex(
+                    module,
+                    PathBuf::from(&self.output_path),
+                )),
+                #[cfg(feature = "wabt")]
+                "wat" => Some(ChiselOutputWriter::Wat(
+                    module,
+                    PathBuf::from(&self.output_path),
+                )),
+                _ => None,
+            };
+            if let Some(writer) = writer {
+                match writer.write() {
+                    Ok(()) => Ok(true),
+                    Err(e) => Err(e),
+                }
+            } else {
+                Err("invalid mode".into())
+            }
+        } else {
+            Ok(false)
+        }
+    }
+}
+
+impl ChiselOutputWriter {
+    /// Serializes and writes the contained module.
+    pub fn write(self) -> Result<(), Box<dyn Error>> {
+        let ret = match self {
+            ChiselOutputWriter::Bin(module, path) => {
+                if *path == PathBuf::from("/dev/stdout") || *path == PathBuf::from("/dev/stderr") {
+                    return Err("cannot write raw binary to a standard stream".into());
```

Comment:
Why is this disallowed?

Reply:
Shits out escape codes you might not want, kinda like reading from /dev/random